config/yaml.go

```diff
@@ -29,34 +29,71 @@ import (
 	"strings"
 
 	"gopkg.in/yaml.v2"
+	"strconv"
 )
 
 type yamlConfigProvider struct {
-	roots  []*yamlNode
+	root   *yamlNode
 	vCache map[string]Value
 }
 
 var _ Provider = &yamlConfigProvider{}
 
 func newYAMLProviderCore(files ...io.ReadCloser) Provider {
-	roots := make([]*yamlNode, len(files))
-
-	for n, v := range files {
-		if root, err := newyamlNode(v); err != nil {
+	var root interface{} = make(map[interface{}]interface{})
+	for _, v := range files {
+		curr := make(map[interface{}]interface{})
+		if err := unmarshalYamlValue(v, &curr); err != nil {
 			panic(err)
-		} else {
-			roots[n] = root
 		}
+
+		root = mergeMaps(root, curr)
 	}
 
 	return &yamlConfigProvider{
-		roots:  roots,
+		root: &yamlNode{
+			nodeType: objectNode,
+			key:      Root,
+			value:    root},
 		vCache: make(map[string]Value),
 	}
 }
 
-// NewYAMLProviderFromFiles creates a configration provider from a set of YAML file
-// names
+// We need to have a custom merge map because yamlV2 doesn't unmarshal `map[interface{}]map[interface{}]interface{}`
+// as we expect: it will replace second level maps with new maps on each unmarshal call, instead of merging them.
+func mergeMaps(dst interface{}, src interface{}) interface{} {
+	if dst == nil {
+		panic("Destination node is nil")
+	}
+
+	if src == nil {
+		return src
+	}
+
+	switch s := src.(type) {
+	case map[interface{}]interface{}:
+		dstMap, ok := dst.(map[interface{}]interface{})
+		if !ok {
+			panic(fmt.Sprintf("Expected map[interface{}]interface{}, actual: %+v", dstMap))
+		}
+
+		for k, v := range s {
+			oldVal := dstMap[k]
+			if oldVal == nil {
+				dstMap[k] = v
+			} else {
+				dstMap[k] = mergeMaps(oldVal, v)
+			}
+		}
+	default:
+		dst = src
+	}
+
+	return dst
```

Comment:
Maybe doesn't matter depending on how often this is called, but should we just modify the maps in place instead of allocating and returning a new one?